### PR TITLE
[stable/kong] introduce Kong ingress controller

### PR DESCRIPTION
**Pull request by hbagdi**
#### What this PR does / why we need it:

Deploy Kong as your Ingress controller!

This is developed by @merqurio and I'm merely making some doc changes and opening this PR.
#7350 is superseded by this.

### CODE REVIEW

**Comment by hbagdi** [comment]
@cpanato Could you help me figure out why this E2E fails?
The same migration container runs fine in previous deployments but in the third one it fails even though the process inside the container exits fine.

**Comment by cpanato** [comment]
@hbagdi I will take a look on that tomorrow, if is that ok for you

**Comment by hbagdi** [comment]
@cpanato Thank you for the prompt response. Tomorrow works, thanks!

**Comment by merqurio** [comment]
Thanks @hbagdi I really appreciate this. I'm full of work and didn't found time to commit to this. I had to create a fork due to [this issue][https://github.com/helm/helm/issues/4490] and didn't got the time to go back to this fork.

File: stable/kong/templates/crd-kongconsumer.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{{- if .Values.ingressController.enabled -}}
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
```
Comment (cpanato) [inline stable/kong/templates/crd-kongconsumer.yaml:4]: can you add the 
```
annotations:
    "helm.sh/hook": crd-install
```

in all crds ?

File: stable/kong/values.yaml [hunk 2]
```diff
@@ -152,3 +152,35 @@ postgresql:
   postgresDatabase: kong
   service:
     port: 5432
+  persistence:
+    enabled: false
```
Comment (cpanato) [inline stable/kong/values.yaml]: Persistence should be set to true by default

**Comment by hbagdi** [comment]
@cpanato It seems like now we've an issue with (un)installing CRDs on the CI cluster.

cc @scottrigby @unguiculus

**Comment by unguiculus** [comment]
I deleted the CRDs manually. Can you make sure the chart removes them when it is uninstalled?

File: stable/kong/templates/controller-cluster-role.yaml [hunk 3]
```diff
@@ -0,0 +1,68 @@
+{{- if .Values.ingressController.enabled -}}
```
Comment (unguiculus) [inline stable/kong/templates/controller-cluster-role.yaml]: Please follow our RBAC best practices:
https://github.com/helm/helm/blob/master/docs/chart_best_practices/rbac.md

File: stable/kong/templates/controller-deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+kind: Deployment
+metadata:
+  name: "{{ template "kong.fullname" . }}-controller"
+  labels:
+    app: "{{ template "kong.name" . }}"
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+    component: "app"
+  namespace: {{ .Release.Namespace }}
```
Comment (unguiculus) [inline stable/kong/templates/controller-deployment.yaml]: It is unnecessary to specify the namespace on resources. This is handled by Helm automatically.

File: stable/kong/templates/controller-deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+spec:
+  replicas: {{ .Values.ingressController.replicaCount }}
+  selector:
+    matchLabels:
+      app: {{ template "kong.name" . }}
+      release: {{ .Release.Name }}
```
Comment (unguiculus) [inline stable/kong/templates/controller-deployment.yaml:17]: Selector needs `component` label.

File: stable/kong/templates/controller-deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+  template:
+    metadata:
+    {{- if .Values.podAnnotations }}
+      annotations:
+{{ toYaml .Values.podAnnotations | indent 8 }}
+    {{- end }}
+      labels:
+        app: {{ template "kong.name" . }}
+        release: {{ .Release.Name }}
```
Comment (unguiculus) [inline stable/kong/templates/controller-deployment.yaml:27]: Selector needs `component` label.

File: stable/kong/templates/controller-deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+apiVersion: apps/v1beta2
+kind: Deployment
+metadata:
+  name: "{{ template "kong.fullname" . }}-controller"
+  labels:
+    app: "{{ template "kong.name" . }}"
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+    component: "app"
```
Comment (unguiculus) [inline stable/kong/templates/controller-deployment.yaml]: The Kong deployment has the same component label. You need a different one here.

**Review by unguiculus — CHANGES_REQUESTED** [review]
The failing CI is caused by issues with the component label on the ingress controller.

**Comment by hbagdi** [comment]
@unguiculus I'll work on cleaning up RBAC stuff but I think I'll need help from you to clean up the CRDs once more.

**Comment by hbagdi** [comment]
@unguiculus  Testing locally, the `crd-install` hook seems to have a problem.

It doesn't actually install the CRD.

Running the following:
```
helm install --dry-run --debug . \
    --set ingressController.enabled=true \
    --namespace yolo \
    --set admin.useTLS=false \
    --set readinessProbe.httpGet.scheme=HTTP \
    --set livenessProbe.httpGet.scheme=HTTP
```

give the following output:
https://gist.github.com/hbagdi/fb9edd4c4657670cc12bbf77e141df5a

**Comment by hbagdi** [comment]
The tests should now pass once the CRDs are deleted manually once from the test infra cluster.
We're not using a `hook` for the chart since we don't use the any object of that CR type.

**Comment by hbagdi** [comment]
@unguiculus Really appreciate you jumping in on this one and helping out troubleshoot the CI issue. Could you review this once more? 
Thank you!

**Review by shashiranjan84 — APPROVED** [review]
/lgtm

**Comment by hbagdi** [comment]
@unguiculus Gentle ping for an approval on this PR.

**Comment by merqurio** [comment]
Thanks @hbagdi  for this

**Comment by hbagdi** [comment]
@merqurio @cpanato @unguiculus @shashiranjan84 Thank you for helping get this through!

**Comment by joe1chen** [comment]
@hbagdi Great work on this! I've been following this issue and testing it the past few days. 

I found a pretty severe bug with this pull though.  It will break the kong proxy because a bug is introduced where traffic is occasionally sent to the ingress controller pod instead of the proxy pods.  I kept on getting intermittent connection refused errors until I found the issue is a result of the services not having a specific enough selector.

In service-kong-admin.yaml:
Change:
```
  selector:
    app: {{ template "kong.name" . }}
    release: {{ .Release.Name }}
```
to
```
  selector:
    app: {{ template "kong.name" . }}
    release: {{ .Release.Name }}
    component: app
```

And the same in service-kong-proxy.yaml:
Change:
```
  selector:
    app: {{ template "kong.name" . }}
    release: {{ .Release.Name }}
```
to:
```
  selector:
    app: {{ template "kong.name" . }}
    release: {{ .Release.Name }}
    component: app
```

Normally I would fork and open a pull request, but the last time I tried to contribute to helm, I had issues getting my environment set up to sign the contributor agreement.  So in the interests of time and because this is a severe bug, I'm hoping you can open the bug for me and fix it.

I also have a few other enhancements to contribute, such as support for TLS, which is working well with cert-manager and LetsEncrypt.

**Pull request merged**